### ISSUE TITLE
Sad, No clear Documentaion

### ISSUE THREAD

**Issue by afzl-wtu** [mholt/caddy-l4#254]
It is really very sad that after spending one hour on this app and compiling my caddy with this app. I am still at no where. Whenever I try to copy code provided in README.md file for IMAP setup i get this:
```
docker exec caddy sh -c "cd /etc/caddy && caddy fmt --overwrite && caddy reload"
{"level":"info","ts":1728466368.791589,"msg":"using adjacent Caddyfile"}
Error: adapting config using caddyfile: Caddyfile:14: unrecognized directive: layer4
```

I do not know where to put your provided code:
```
layer4 {
        0.0.0.0:993 {
            route {
                tls
                proxy {
                    proxy_protocol v1
                    upstream localhost:143
                }
            }
        }
}
```
I  want caddy to handle my caddy generated ssl certificate at port 993 for Imap and then proxy unencrypted connection to port 143

I want something like this in my caddyFile:
```
# my mail domain with caddy auto ssl
mail.mydomain.com:993 {
     reverse_proxy 127.0.0.1:143
}
```

But I am still at nowhere.

**Comment by afzl-wtu** [comment]
Finally I got it make work by placing
```
{
        layer4 {
                127.0.0.1:993 {
                        route {
                                tls
                                proxy {
                                        proxy_protocol v1
                                        upstream 127.0.0.1:143
                                }
                        }
                }
                127.0.0.1:465 {
                        route {
                                tls
                                proxy {
                                        proxy_protocol v1
                                        upstream 127.0.0.1:587
                                }
                        }
                }
        }
}
```


At top of my Caddyfile.
But I now have a question. any device that try to connect to my server, will it get an auto-generated letsencrypt ssl certificate for mail.mydomain.com or it will get self signed certificate. If it get self signed certificate is it a security risk or how to assign letsencrypt certificate?

**Comment by ArcCal** [comment]
If you like a particular plugin in Caddy, in addition to reading the documentation, it's a good idea to closely follow the discussions in the issues section. Many use cases and potential pitfalls are discussed there.

**Comment by ArcCal** [comment]
The documentation issues with Caddy plugins have been persistent for a long time, and the discussions in the issues section are an essential source of information or means of obtaining it.

**Comment by vnxme** [comment]
@afzl-wtu You are right it would be great to have more documentation for all the wonderful features this project has. As fairly mentioned by @ArcCal, issues and PRs are valuable sources of information here, in addition to the comments written in the code. Anyone may also help the project by composing extra docs, examples, etc.

> But I now have a question. any device that try to connect to my server, will it get an auto-generated letsencrypt ssl certificate for mail.mydomain.com or it will get self signed certificate. If it get self signed certificate is it a security risk or how to assign letsencrypt certificate?

The answer to your question really depends on what you have in your config. The layer4 module doesn't generate any certificates itself. When `tls` handler is used, it tries to find the best available certificate for the requested domain name and, if found, terminates TLS. And the certificates are those generally available to/generated by Caddy.

You may influence certificate selection/generation by specifying [TLS options](https://caddyserver.com/docs/caddyfile/directives/tls) inside a website block:
```
{
	layer4 {
		:993 {
			...
		}
		:465 {
			...
		}
	}
}

mail.mydomain.com {
	tls ...
}
```

**Comment by mholt** [comment]
Hi, I actually agree, the docs for this module could be greatly improved. I've simply been too busy to prioritize it myself.

Caddyfile support is a relatively new enhancement by the way.

I would gladly accept pull requests to improve the docs!

We could also potentially start a wiki if needed, maybe with examples, to keep the README tidy.

**Comment by polarathene** [comment]
> But I now have a question. any device that try to connect to my server, will it get an auto-generated letsencrypt ssl certificate for `mail.mydomain.com` or it will get self signed certificate.

Just for reference, I covered this in my reproduction example [here](https://github.com/lucaslorentz/caddy-docker-proxy/issues/702#issuecomment-2765242635) (_for SSH on port 22, using TLS via Caddy L4 to route TLS wrapped client connections_).

Some times it can be convenient to have a wildcard provisioned, and then give it a fallback `respond` / `abort` directive for HTTPS connections, while allowing the L4 `tls` to match the cert via TLS SNI 👍 

```
# Global Settings:
{
  auto_https prefer_wildcard
}

*.example.com {
  # Use `respond` instead of `abort` if you're troubleshooting connectivity:
  # respond "hello world"

  # Reject connection when subdomain has no configured container:
  abort
}
```

**NOTE:** `auto_https prefer_wildcard` may not be required in a future version of Caddy (current 2.9.0) as it may become the default setting. This will choose the wildcard certificate instead of provisioning a new one for a subdomain if the wildcard is already provisioned.

Self-signed vs LetsEncrypt depends on the configured site-address and global settings (_use `local_certs` global setting, or `tls internal` site-address setting for example to force self-signed_).

With the wildcard cert approach, if you have various L4 config entries, it may be a bit nicer than maintaining individual subdomains. For LetsEncrypt it'll complicate provisioning a little bit as you'll need a plugin for your DNS provider to handle the DNS ACME challenge instead.

> If it get self signed certificate is it a security risk or how to assign letsencrypt certificate?

The security risk concern is only regarding trust of the private CA cert vs a public CA cert (_LetsEncrypt, that your systems already trust_):
- If your connecting client does not trust it, but enforces TLS verification the connection will fail.
- If your connecting client does trust the cert, it should be just as reliable as LetsEncrypt provided the CA private key is not compromised.

---

With Caddy Docker Proxy, you could have a similar config like that above `Caddyfile` snippet (_see earlier link for full config reference_):

```yaml
services:
  reverse-proxy:
    # Global options first, then a separate site-address block:
    labels:
      # Prefer using a wildcard certificate when already available instead of provisioning a new cert:
      caddy.auto_https: prefer_wildcard
      # Provision a wildcard cert:
      caddy_1: "*.example.internal"
      # Double quoting required for preserving quotes (YAML string => Caddy string where inner quotes required)
      caddy_1.respond: '"Hello World"'
```
